From 178d1e149477c2d55568a533829573021f65c554 Mon Sep 17 00:00:00 2001
From: Samuel Dionne-Riel <samuel@dionne-riel.com>
Date: Sat, 12 Feb 2022 13:53:12 -0500
Subject: [PATCH] meson: Prefer internal boot methods first

---
 include/configs/meson64.h | 11 +++++------
 1 file changed, 5 insertions(+), 6 deletions(-)

diff --git a/include/configs/meson64.h b/include/configs/meson64.h
index 196e58ed9a3..78e2a477727 100644
--- a/include/configs/meson64.h
+++ b/include/configs/meson64.h
@@ -68,15 +68,14 @@
 
 #ifndef BOOT_TARGET_DEVICES
 #define BOOT_TARGET_DEVICES(func) \
-	func(ROMUSB, romusb, na)  \
-	func(MMC, mmc, 0) \
-	func(MMC, mmc, 1) \
-	func(MMC, mmc, 2) \
-	BOOT_TARGET_DEVICES_USB(func) \
+	func(MMC, mmc, 2) /* eMMC */ \
 	BOOT_TARGET_NVME(func) \
 	BOOT_TARGET_SCSI(func) \
+	func(MMC, mmc, 1) /* SD */ \
+	BOOT_TARGET_DEVICES_USB(func) \
 	func(PXE, pxe, na) \
-	func(DHCP, dhcp, na)
+	func(DHCP, dhcp, na) \
+	func(ROMUSB, romusb, na)
 #endif
 
 #include <config_distro_bootcmd.h>
-- 
2.38.0

