From 845d1ef559cdca814eee4fcff215a18b22b18909 Mon Sep 17 00:00:00 2001
From: Samuel Dionne-Riel <samuel@dionne-riel.com>
Date: Sat, 12 Feb 2022 13:53:12 -0500
Subject: [PATCH] meson: Prefer eMMC to SD card boot

---
 include/configs/meson64.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/include/configs/meson64.h b/include/configs/meson64.h
index f9bb0240d2c..a5753b14c5e 100644
--- a/include/configs/meson64.h
+++ b/include/configs/meson64.h
@@ -69,8 +69,8 @@
 #ifndef BOOT_TARGET_DEVICES
 #define BOOT_TARGET_DEVICES(func) \
 	func(ROMUSB, romusb, na)  \
-	func(MMC, mmc, 0) \
 	func(MMC, mmc, 1) \
+	func(MMC, mmc, 0) \
 	func(MMC, mmc, 2) \
 	BOOT_TARGET_DEVICES_USB(func) \
 	BOOT_TARGET_NVME(func) \
-- 
2.34.0

