From 2c78039ec062e3c537822e431109e01a845ad7ea Mon Sep 17 00:00:00 2001
From: Da Xue <da@libre.computer>
Date: Fri, 18 Jun 2021 19:48:10 -0400
Subject: [PATCH] autoboot: show menu only on menu key

Origin: https://github.com/libre-computer-project/libretech-u-boot/commit/dd14366c023ffc35500cf241643f6733bd39e564
---
 common/autoboot.c | 29 ++++++++++++++---------------
 1 file changed, 14 insertions(+), 15 deletions(-)

diff --git a/common/autoboot.c b/common/autoboot.c
index db7cdf7ade8..f4cedffdf9d 100644
--- a/common/autoboot.c
+++ b/common/autoboot.c
@@ -372,11 +372,13 @@ static int abortboot_single_key(int bootdelay)
 	 * Check if key already pressed
 	 */
 	if (tstc()) {	/* we got a key press	*/
-		getchar();	/* consume input	*/
-		puts(ANSI_CLEAR_LINE);
-		printf(ANSI_CURSOR_COLUMN, 1);
-		printf(CONFIG_AUTOBOOT_PROMPT, 0);
-		abort = 1;	/* don't auto boot	*/
+		menukey = getchar();
+		if (!IS_ENABLED(CONFIG_AUTOBOOT_USE_MENUKEY) || menukey == AUTOBOOT_MENUKEY){
+			puts(ANSI_CLEAR_LINE);
+			printf(ANSI_CURSOR_COLUMN, 1);
+			printf(CONFIG_AUTOBOOT_PROMPT, 0);
+			abort = 1;	/* don't auto boot	*/
+		}
 	}
 
 	while ((bootdelay > 0) && (!abort)) {
@@ -385,14 +387,12 @@ static int abortboot_single_key(int bootdelay)
 		ts = get_timer(0);
 		do {
 			if (tstc()) {	/* we got a key press	*/
-				int key;
-
-				abort  = 1;	/* don't auto boot	*/
-				bootdelay = 0;	/* no more delay	*/
-				key = getchar();/* consume input	*/
-				if (IS_ENABLED(CONFIG_AUTOBOOT_USE_MENUKEY))
-					menukey = key;
-				break;
+				menukey = getchar();
+				if (!IS_ENABLED(CONFIG_AUTOBOOT_USE_MENUKEY) || menukey == AUTOBOOT_MENUKEY){
+					abort  = 1;	/* don't auto boot	*/
+					bootdelay = 0;	/* no more delay	*/
+					break;
+				}
 			}
 			udelay(10000);
 		} while (!abort && get_timer(ts) < 1000);
@@ -502,8 +502,7 @@ void autoboot_command(const char *s)
 			disable_ctrlc(prev);	/* restore Ctrl-C checking */
 	}
 
-	if (IS_ENABLED(CONFIG_AUTOBOOT_USE_MENUKEY) &&
-	    menukey == AUTOBOOT_MENUKEY) {
+	if (IS_ENABLED(CONFIG_AUTOBOOT_USE_MENUKEY)){
 		s = env_get("menucmd");
 		if (s)
 			run_command_list(s, -1, 0);
-- 
2.35.1

